### PR TITLE
feat: email confirmation on change

### LINKED ISSUE #8856: [CAL-1670] No verification required, when changing email
### Issue Summary

No verification required, when changing email.

### Steps to Reproduce

1. Went to profile section in settings
2. Changed email

A verification mail can be sent to confirm that user has not misspelled his/her email. And user can retrieve his/her account in case he/she forgot password.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1670](https://linear.app/calcom/issue/CAL-1670/no-verification-required-when-changing-email)</sub>

### PR DESCRIPTION
**Pull request by sean-brydon**
Show old and new email in email confirmation box 

Fixes: #8856 

https://github.com/calcom/cal.com/assets/55134778/5314db2e-ff26-4b4b-9e9d-37758a26f81c

### CODE REVIEW

**Comment by sean-brydon** [comment]
@itsberkaya @ciaranha  maybe we can come up with a proper design for this in the future - i think this is fine for now let me know if you want any changes here.

**Review by keithwillcode — COMMENTED** [review]
@sean-brydon the [issue](https://github.com/calcom/cal.com/issues/8856) mentioned as being fixed here says that no confirmation email is sent to confirm the new email is correct. Is there more to be added to this PR or was it decided that won’t be implemented?

**Review by Udit-takkar — COMMENTED** [review]
When Changing email it would be better to show verify email dialog (like on booking page). 

Currently after you change your email the emailVerified is set to null and you see banner only after some time or reloading and no email is sent to verify email. you have to click on 'Resend Email' to actually send the email

Screenshot 2024-01-29 at 4 18 36 PM

**Comment by sean-brydon** [comment]
> When Changing email it would be better to show verify email dialog (like on booking page).
> 
> Currently after you change your email the emailVerified is set to null and you see banner only after some time or reloading and no email is sent to verify email. you have to click on 'Resend Email' to actually send the email
> 
> Screenshot 2024-01-29 at 4 18 36 PM

Yeah in talks with @ciaranha about this as a improvement to this PR - great shout udit this is what we will be going with 

will return to draft and update accordingly

File: apps/web/pages/auth/verify-email-change.tsx
Comment (sean-brydon) [inline apps/web/pages/auth/verify-email-change.tsx:1]: Now this file is a bit odd. We can't easily update the session from the serverside. So we provide hacky approach to do it for us. 

Nothing unsafe about this - update session is a public endpoint that gets updated in the `trigger==="update"` side of the jwt callback inside `next-auth-options.ts` 

This is how we handle this for none verified emails on the FE too.

File: apps/web/pages/auth/verify-email-change.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      }
+    }
+    // We only need this to run on inital mount. These props can't and won't change due to it being fetched serveside.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  return (
+    <div className="flex h-screen w-full items-center justify-center">
+      <Toaster position="bottom-right" />
+    </div>
```
Comment (sean-brydon) [inline apps/web/pages/auth/verify-email-change.tsx:45]: Need to render a toaster here as it doesnt exist before the navigation happens.

File: apps/web/pages/auth/verify-email-change.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    return {
+      notFound: true,
+    };
+  }
+
+  const params = new URLSearchParams({
+    token,
+  });
+
+  // Fetch data based on `slug` from your API or any data source
```
Comment (sean-brydon) [inline apps/web/pages/auth/verify-email-change.tsx:66]:
```suggestion
```

File: apps/web/pages/auth/verify-email-change.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      <Toaster position="bottom-right" />
+    </div>
+  );
+}
+
+const tokenSchema = z.object({
+  token: z.string(),
+});
+
+export async function getServerSideProps(context: GetServerSidePropsContext) {
```
Comment (sean-brydon) [inline apps/web/pages/auth/verify-email-change.tsx:53]: Fetch serverside with the token passed in as props

**Review by Udit-takkar — COMMENTED** [review]
https://github.com/calcom/cal.com/assets/53316345/d4f2ae65-e77c-41a5-94de-e35e33681c1c

When I tried to change my email then I wasn't logged out immediately but only when i tried to reload or go to other page. The toast message says the email will be updated later but the UI displays the new email on settings/my-account/profile.

**Comment by sean-brydon** [comment]
> Screen.Recording.2024-02-07.at.12.40.55.PM.mov 
> 
> When I tried to change my email then I wasn't logged out immediately but only when i tried to reload or go to other page. The toast message says the email will be updated later but the UI displays the new email on settings/my-account/profile.

Oh I know what’s happening here - can you test the rest of the flow will push a fix for this now

**Review by Udit-takkar — COMMENTED** [review]
When I am logged in with Gmail (Idenitity provider) and try to change my email, I got logged out and got an email to set password and after logging in, I see email has not been verified which should have been verified as the password reset url was sent to email

Screenshot 2024-02-07 at 1 32 19 PM

**Review by Udit-takkar — COMMENTED** [review]
If i try to change the email to one that already exists in the DB then there is no message like 'This email has been taken' and verify email is sent but the link doesn't work as the email already exist for other user

Screenshot 2024-02-07 at 1 54 25 PM

**Comment by sean-brydon** [comment]
> When I am logged in with Gmail (Idenitity provider) and try to change my email, I got logged out and got an email to set password and after logging in, I see email has not been verified which should have been verified as the password reset url was sent to email
> 
> Screenshot 2024-02-07 at 1 32 19 PM

Hmm - i thought we had tests for this so expected it to fail if this borked. Will write tests and fix

**Review by Udit-takkar — APPROVED** [review]
LGTM 🥇 💯 . Thanks for Fixing all the bugs and adding tests

**Pull request merged**